modules/managed-identity/user-assigned-identity/main.bicep

```diff
@@ -0,0 +1,57 @@
+@description('Optional. Name of the User Assigned Identity.')
+param name string = guid(resourceGroup().id)
+
+@description('Optional. Location for all resources.')
+param location string = resourceGroup().location
+
+@allowed([
+  'CanNotDelete'
+  'NotSpecified'
+  'ReadOnly'
+])
+@description('Optional. Specify the type of lock.')
+param lock string = 'NotSpecified'
+
+@description('Optional. Array of role assignment objects that contain the \'roleDefinitionIdOrName\' and \'principalId\' to define RBAC role assignments on this resource. In the roleDefinitionIdOrName attribute, you can provide either the display name of the role definition, or its fully qualified ID in the following format: \'/providers/Microsoft.Authorization/roleDefinitions/c2f4ef07-c644-48eb-af81-4b1b4947fb11\'')
+param roleAssignments array = []
+
+@description('Optional. Tags of the resource.')
+param tags object = {}
+
+resource userMsi 'Microsoft.ManagedIdentity/userAssignedIdentities@2018-11-30' = {
```

Comment:
What is this module helping abstract in this particular case? There are no special restrictions on managed identity resource names, and there are no additional properties to set that are complicated. It seems the "raw" resource is simpler than using this module since with the module I would be exposed to extra parameters for locking and rbac. 

I guess I would not need to set the name & location myself if I use the module, but that is very basic and something we would want all bicep developers to be able to do.

Reply:
In this particular case, there are no child-modules, so that leaves us with the extension resources - that is RBAC that you can deploy together with the MSI.

Reply:
I guess I am still ok checking this in for now, but TBH, I'm growing more concerned about this pattern when I see modules like this. 

I'd prefer customers learn how to do extension resources once, then be able to repeat it as they need it. Hiding it inside of a module allows the module to be more functional at the cost of user education. Would you be willing to submit all CARML modules without those extension resources? 

If we have those, then you could in theory make an "L2" module that combines this extension-less modules with rbac + locks. I recognize that is more work, but feels like the right layering. Thoughts?

Reply:
I agree we shold have a fundamental discussion about some of the patterns we apply (i.e. extension resources, what we optimize for in CARML, and also what vice-versa what you want to go for with the bicep registry). Probably not in a comment section in GitHub though as this would turn into an essay ;)

We could publish without the extension resources (that leaves the child-resources as the biggest addition), but I'd argue that it wouldn't make sense & would not be worthwhile doing as we're then clearly engineering into 2 different directions - making it increasingly difficult & time intensive to re-engineer the modules to account for the registry's intended module design.

Reply:
I had a similar thought/reaction as @alex-frankel on this one - on why we're bundling things together like this (which seems the pattern in CARML now that I spend more time with it).  My first confusion was that I expected the roleAssignment to give the MSI access to something - that's the common pattern.  Create the MSI and then give it perms to do something.  Where this seems to be just giving people access to an MSI that doesn't have access to anything (since it was just created).  I "get" the pattern of -- create resources, let someone access them -- but for MSIs this seems less useful without being able to state what the MSI will do (i.e. have access to).

The other thought is whether we want to create new ways for users to do common/basic things...  IOW, do we want another/different way to create a resource and give access to it?  It is a tricky line, if you're familiar with ARM and you understand the simplification, it can be helpful but if you're new it will be confusing because every doc we have will talk about the native way of doing things, not this one.

Interesting (longer term) discussion of whether some of these scenarios (e.g. locks) should be simplified in the platform.